trace/obfuscate.go

```diff
@@ -1,40 +1,204 @@
+		case *pb.HTTPMeta_Cookie:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Header:
+			header := httpMeta.GetHeader()
+			key = header.Key
+		case *pb.HTTPMeta_Query:
+			queryParam := httpMeta.GetQuery()
+			key = queryParam.Key
+		}
+		// Check if the key is in the list of keys to obfuscate.
+		if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(key)) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	primitive := d.GetPrimitive()
+	if primitive == nil {
+		// Not a primitive, regex will be applied to primitive string values only.
+		return Continue
+	}
+
+	stringValue := primitive.GetStringValue()
+	if stringValue == nil {
+		// Not a string, regex will be applied to string values only.
+		return Continue
+	}
+
+	for pattern, _ := range s.obfuscationOptions.SensitiveDataValuePatterns {
+		if pattern.MatchString(stringValue.Value) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) EnterHTTPBody(self interface{}, ctx vis.SpecVisitorContext, b *pb.HTTPBody) Cont {
+	node, _ := ctx.GetInnermostNode(reflect.TypeOf((*pb.Data)(nil)))
+	data := node.(*pb.Data)
+	if data == nil {
+		return Continue
+	}
+
+	// Traverse the data node and redact sensitive information.
+	s.traverseAndRedactSensitiveInfo(ctx, data)
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) traverseAndRedactSensitiveInfo(ctx vis.SpecVisitorContext, data *pb.Data) {
+	switch val := data.Value.(type) {
+	case *pb.Data_Struct:
+		// Traverse the struct's fields and redact sensitive information based on key.
+		structData := val.Struct
+		for fieldName, fieldData := range structData.GetFields() {
+			// If field Data is primitive or primitiveList, then check it's key and redact if needed
+			if spec_util.IsPrimitive(fieldData) || spec_util.IsPrimitiveList(fieldData) {
+				if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(fieldName)) {
+					ObfuscatePrimitiveWithRedactedString(fieldData)
+				}
+			} else {
+				// Recursively traverse the struct's fields.
+				childCtx := ctx.EnterStruct(structData, fieldName).(vis.SpecVisitorContext)
+				s.traverseAndRedactSensitiveInfo(childCtx, fieldData)
+			}
+		}
+	case *pb.Data_List:
+		listData := val.List
+		if items := listData.GetElems(); items != nil {
+			for i, itemData := range items {
+				// Step through each item in the list and traverse the data node.
+				// This list will never be primitive list, as primitive list is handled above.
+				childCtx := ctx.EnterArray(listData, i).(vis.SpecVisitorContext)
+				s.traverseAndRedactSensitiveInfo(childCtx, itemData)
+			}
+		}
+	}
+}
+
+// Obfuscate the given primitive data with REDACTED string.
+// In case of a primitive value repalce with REDACTED string and in case of a primitive list replace
+// with a new list having single element, i.e., REDACTED string, to maintain the list structure.
```

Comment:
I am not really sure why the same function is handling both Primitives and Lists.  What do we get from combining them in the same function?

If lists of primitives are handled differently, then I think that should be in the function above where we know it's a list already?